File: experiment-14/runs/language=rust_model=opus-4.8-fast_prompt=ATDD/rep1/src/mcp.rs
//! Minimal MCP (Model Context Protocol) server over a stdio JSON-RPC 2.0
//! transport using newline-delimited messages.
//!
//! Implements the handshake (`initialize` / `notifications/initialized`),
//! `tools/list`, and `tools/call`. Each tool returns both `content` (text the
//! LLM reads) and `structuredContent` (machine-readable result).

use serde_json::{json, Value};

use crate::data::DataStore;
use crate::tools::{self, ToolOutput};

pub const PROTOCOL_VERSION: &str = "2024-11-05";
pub const SERVER_NAME: &str = "brazilian-soccer-mcp";
pub const SERVER_VERSION: &str = "0.1.0";

type ToolFn = fn(&DataStore, &Value) -> Result<ToolOutput, String>;

struct ToolDef {
    name: &'static str,
    description: &'static str,
    input_schema: Value,
    func: ToolFn,
}

fn string_prop(desc: &str) -> Value {
    json!({ "type": "string", "description": desc })
}
fn int_prop(desc: &str) -> Value {
    json!({ "type": "integer", "description": desc })
}

fn tool_defs() -> Vec<ToolDef> {
    vec![
        ToolDef {
            name: "search_matches",
            description: "Find matches by team, opponent, competition, season, or date range. \
                          When both 'team' and 'opponent' are given, a head-to-head summary is included.",
            input_schema: json!({
                "type": "object",
                "properties": {
                    "team": string_prop("A team to find matches for (home or away). Name variants like 'Palmeiras' or 'Palmeiras-SP' both work."),
                    "opponent": string_prop("Restrict to matches against this opponent."),
                    "home_team": string_prop("Restrict to matches where this team played at home."),
                    "away_team": string_prop("Restrict to matches where this team played away."),
                    "competition": string_prop("Brasileirão, Copa do Brasil, or Copa Libertadores."),
                    "season": int_prop("Season year, e.g. 2019."),
                    "start_date": string_prop("Inclusive ISO start date (YYYY-MM-DD)."),
                    "end_date": string_prop("Inclusive ISO end date (YYYY-MM-DD)."),
                    "limit": int_prop("Maximum matches to return (default 50).")
                }
            }),
            func: tools::search_matches,
        },
        ToolDef {
            name: "team_record",
            description: "Get a team's win/draw/loss record, goals, points and win rate, \
                          optionally filtered by competition, season, and venue (home/away/all).",
            input_schema: json!({
                "type": "object",
                "properties": {
                    "team": string_prop("Team name (required)."),
                    "competition": string_prop("Optional competition filter."),
                    "season": int_prop("Optional season year filter."),
                    "venue": json!({ "type": "string", "enum": ["home", "away", "all"], "description": "Restrict to home or away games (default all)." })
                },
                "required": ["team"]
            }),
            func: tools::team_record,
        },
        ToolDef {
            name: "head_to_head",
            description: "Compare two teams head-to-head: total meetings, each side's wins, draws, goals, and the match list.",
            input_schema: json!({
                "type": "object",
                "properties": {
                    "team_a": string_prop("First team (required)."),
                    "team_b": string_prop("Second team (required)."),
                    "competition": string_prop("Optional competition filter.")
                },
                "required": ["team_a", "team_b"]
            }),
            func: tools::head_to_head,
        },
        ToolDef {
            name: "search_players",
            description: "Search FIFA players by name, nationality, club, position and minimum overall rating; \
                          sortable by overall, potential or age.",
            input_schema: json!({
                "type": "object",
                "properties": {
                    "name": string_prop("Full or partial player name."),
                    "nationality": string_prop("Nationality, e.g. 'Brazil'."),
                    "club": string_prop("Club name (partial match)."),
                    "position": string_prop("Playing position, e.g. 'ST', 'GK'."),
                    "min_overall": int_prop("Minimum FIFA overall rating."),
                    "sort_by": json!({ "type": "string", "enum": ["overall", "potential", "age"], "description": "Sort key (default overall, descending)." }),
                    "limit": int_prop("Maximum players to return (default 25).")
                }
            }),
            func: tools::search_players,
        },
        ToolDef {
            name: "league_standings",
            description: "Compute the final league table for a competition and season from match results \
                          (rank, points, played, W/D/L, goals for/against, goal difference).",
            input_schema: json!({
                "type": "object",
                "properties": {
                    "competition": string_prop("Competition (default Brasileirão)."),
                    "season": int_prop("Season year (required), e.g. 2019.")
                },
                "required": ["season"]
            }),
            func: tools::league_standings,
        },
        ToolDef {
            name: "competition_stats",
            description: "Aggregate statistics for a competition/season: match count, average goals per match, \
                          home/away/draw split, home win rate, and the biggest wins.",
            input_schema: json!({
                "type": "object",
                "properties": {
                    "competition": string_prop("Optional competition filter."),
                    "season": int_prop("Optional season year filter.")
                }
            }),
            func: tools::competition_stats,
        },
        ToolDef {
            name: "list_competitions",
            description: "List all competitions available in the loaded data with match counts and season ranges.",
            input_schema: json!({ "type": "object", "properties": {} }),
            func: tools::list_competitions,
        },
    ]
}

pub struct Server {
    store: DataStore,
    tools: Vec<ToolDef>,
}

impl Server {
    pub fn new(store: DataStore) -> Self {
        Server { store, tools: tool_defs() }
    }

    /// Handle one JSON-RPC message. Returns `Some(response)` for requests and
    /// `None` for notifications (which receive no reply).
    pub fn handle(&self, msg: &Value) -> Option<Value> {
        let id = msg.get("id").cloned();
        let method = msg.get("method").and_then(|m| m.as_str()).unwrap_or("");
        let params = msg.get("params").cloned().unwrap_or(json!({}));

        // Notifications carry no id and expect no response.
        if id.is_none() {
            return None;
        }
        let id = id.unwrap();

        let result = match method {
            "initialize" => Ok(json!({
                "protocolVersion": PROTOCOL_VERSION,
                "capabilities": { "tools": {} },
                "serverInfo": { "name": SERVER_NAME, "version": SERVER_VERSION }
            })),
            "ping" => Ok(json!({})),
            "tools/list" => Ok(self.list_tools()),
            "tools/call" => self.call_tool(&params),
            other => Err((-32601, format!("method not found: {other}"))),
        };

        Some(match result {
            Ok(value) => json!({ "jsonrpc": "2.0", "id": id, "result": value }),
            Err((code, message)) => json!({
                "jsonrpc": "2.0",
                "id": id,
                "error": { "code": code, "message": message }
            }),
        })
    }

    fn list_tools(&self) -> Value {
        let tools: Vec<Value> = self
            .tools
            .iter()
            .map(|t| {
                json!({
                    "name": t.name,
                    "description": t.description,
                    "inputSchema": t.input_schema,
                })
            })
            .collect();
        json!({ "tools": tools })
    }

    fn call_tool(&self, params: &Value) -> Result<Value, (i64, String)> {
        let name = params
            .get("name")
            .and_then(|n| n.as_str())
            .ok_or((-32602, "missing tool name".to_string()))?;
        let args = params.get("arguments").cloned().unwrap_or(json!({}));

        let tool = self
            .tools
            .iter()
            .find(|t| t.name == name)
            .ok_or((-32602, format!("unknown tool: {name}")))?;

        // Tool-level failures are reported as a successful JSON-RPC result with
        // isError=true, per the MCP convention, so the LLM can read the message.
        match (tool.func)(&self.store, &args) {
            Ok(out) => Ok(json!({
                "content": [{ "type": "text", "text": out.text }],
                "structuredContent": out.structured,
                "isError": false
            })),
            Err(message) => Ok(json!({
                "content": [{ "type": "text", "text": format!("Error: {message}") }],
                "isError": true
            })),
        }
    }
}

File: experiments/adrianco/experiment-14/runs/language=clojure_model=opus-4.8-fast_prompt=neutral/rep1/src/brsoccer/mcp.clj
;; =============================================================================
;; brsoccer.mcp
;;
;; Context:
;;   Implements the Model Context Protocol (https://modelcontextprotocol.io) as a
;;   JSON-RPC 2.0 server over stdio.  An LLM host (Claude Desktop, etc.) launches
;;   this process, performs the `initialize` handshake, lists tools, and calls
;;   them with `tools/call`.  Each tool wraps a brsoccer.query function and
;;   returns BOTH formatted text (for the model to read) and the underlying
;;   structured data (under `structuredContent`) so the model can reason over it.
;;
;;   The protocol plumbing (frame read/write, request dispatch) is deliberately
;;   small and dependency-free; `handle-request` is pure and unit-tested, so the
;;   whole server can be exercised without spawning a process.
;; =============================================================================
(ns brsoccer.mcp
  (:require [clojure.data.json :as json]
            [clojure.string :as str]
            [brsoccer.data :as data]
            [brsoccer.query :as q]
            [brsoccer.format :as fmt])
  (:import [java.io BufferedReader]))

(def protocol-version "2024-11-05")
(def server-info {:name "brazilian-soccer-mcp" :version "1.0.0"})

;; ---------------------------------------------------------------------------
;; Tool definitions
;; ---------------------------------------------------------------------------

(defn- str-prop [desc] {:type "string" :description desc})
(defn- int-prop [desc] {:type "integer" :description desc})

(def tools
  "Each tool: :name :description :inputSchema (JSON Schema) and :handler.
   A handler receives [graph args] (args has keyword keys) and returns
   {:text <string> :data <edn>}."
  [{:name "find_matches"
    :description "Find soccer matches by team, opponent, venue, competition, season or date range. Returns match list plus head-to-head summary when two teams are given."
    :inputSchema {:type "object"
                  :properties {:team (str-prop "Team that played (home or away)")
                               :opponent (str-prop "Restrict to matches against this team")
                               :home (str-prop "Team must have played at home")
                               :away (str-prop "Team must have played away")
                               :competition (str-prop "Competition name substring, e.g. 'Brasileirão', 'Copa do Brasil', 'Libertadores'")
                               :season (int-prop "Season year, e.g. 2019")
                               :from (str-prop "Inclusive start date YYYY-MM-DD")
                               :to (str-prop "Inclusive end date YYYY-MM-DD")
                               :limit (int-prop "Max matches to return (default 20)")}}
    :handler (fn [g a]
               (let [a (update a :limit #(or % 20))
                     ms (q/find-matches g a)]
                 (if (and (:team a) (:opponent a))
                   (let [h (q/head-to-head g {:team-a (:team a) :team-b (:opponent a)
                                              :season (:season a) :competition (:competition a)})]
                     {:text (fmt/head-to-head-block h) :data h})
                   {:text (fmt/matches-block ms :header (format "Found %d match(es):" (count ms)))
                    :data ms})))}

   {:name "team_record"
    :description "Win/draw/loss record, goals and points for a team, optionally filtered by season, competition or venue (home/away)."
    :inputSchema {:type "object"
                  :required ["team"]
                  :properties {:team (str-prop "Team name")
                               :season (int-prop "Season year")
                               :competition (str-prop "Competition name substring")
                               :venue {:type "string" :enum ["home" "away" "all"]
                                       :description "Restrict to home or away matches"}}}
    :handler (fn [g a]
               (let [venue (case (:venue a) "home" :home "away" :away nil)
                     r (q/team-record g (assoc a :venue venue))]
                 {:text (fmt/record-block r) :data r}))}

   {:name "head_to_head"
    :description "Head-to-head record between two teams across the datasets (wins, draws, goals and the matches themselves)."
    :inputSchema {:type "object"
                  :required ["team_a" "team_b"]
                  :properties {:team_a (str-prop "First team")
                               :team_b (str-prop "Second team")
                               :season (int-prop "Restrict to a season")
                               :competition (str-prop "Restrict to a competition")}}
    :handler (fn [g a]
               (let [h (q/head-to-head g {:team-a (:team_a a) :team-b (:team_b a)
                                          :season (:season a) :competition (:competition a)})]
                 {:text (fmt/head-to-head-block h) :data h}))}

   {:name "search_players"
    :description "Search the FIFA player database by name, nationality, club, position and minimum overall rating. Sorted by overall rating descending."
    :inputSchema {:type "object"
                  :properties {:name (str-prop "Player name substring")
                               :nationality (str-prop "Nationality, e.g. 'Brazil'")
                               :club (str-prop "Club name")
                               :position (str-prop "Position substring, e.g. 'ST', 'GK', 'CB'")
                               :min_overall (int-prop "Minimum FIFA overall rating")
                               :limit (int-prop "Max players to return (default 20)")}}
    :handler (fn [g a]
               (let [ps (q/search-players g {:name (:name a) :nationality (:nationality a)
                                             :club (:club a) :position (:position a)
                                             :min-overall (:min_overall a)
                                             :limit (or (:limit a) 20)})]
                 {:text (fmt/players-block ps :header (format "Found %d player(s):" (count ps)))
                  :data ps}))}

   {:name "brazilian_players_by_club"
    :description "Brazilian players grouped by their (Brazilian) club, with player counts and average overall rating."
    :inputSchema {:type "object" :properties {}}
    :handler (fn [g _]
               (let [rows (q/players-by-brazilian-club g)]
                 {:text (str "Brazilian players at Brazilian clubs:\n"
                             (->> rows
                                  (map #(format "- %s: %d players (avg rating: %s)"
                                                (:club %) (:count %) (:avg-overall %)))
                                  (str/join "\n")))
                  :data rows}))}

   {:name "standings"
    :description "Compute a league table (standings) for a competition and season from match results: points, W/D/L, goals for/against, goal difference."
    :inputSchema {:type "object"
                  :required ["competition" "season"]
                  :properties {:competition (str-prop "Competition name substring, e.g. 'Brasileirão'")
                               :season (int-prop "Season year, e.g. 2019")}}
    :handler (fn [g a]
               (let [rows (q/standings g a)]
                 {:text (fmt/standings-block
                          rows :header (format "%s %s standings (calculated from matches):"
                                               (:competition a) (:season a)))
                  :data rows}))}

   {:name "biggest_wins"
    :description "Largest-margin victories, optionally filtered by competition and season."
    :inputSchema {:type "object"
                  :properties {:competition (str-prop "Competition name substring")
                               :season (int-prop "Season year")
                               :limit (int-prop "How many to return (default 10)")}}
    :handler (fn [g a]
               (let [ms (q/biggest-wins g a)]
                 {:text (fmt/matches-block ms :header "Biggest victories in dataset:" :show (or (:limit a) 10))
                  :data ms}))}

   {:name "match_statistics"
    :description "Aggregate statistics over a match set: total/average goals per match and home/away/draw win rates."
    :inputSchema {:type "object"
                  :properties {:competition (str-prop "Competition name substring")
                               :season (int-prop "Season year")}}
    :handler (fn [g a]
               (let [s (q/summary-stats g a)]
                 {:text (fmt/stats-block s) :data s}))}

   {:name "best_record"
    :description "Teams ranked by win rate, optionally for home or away matches only, within a competition/season."
    :inputSchema {:type "object"
                  :properties {:competition (str-prop "Competition name substring")
                               :season (int-prop "Season year")
                               :venue {:type "string" :enum ["home" "away" "all"]}
                               :limit (int-prop "How many teams (default 10)")}}
    :handler (fn [g a]
               (let [venue (case (:venue a) "home" :home "away" :away nil)
                     rows (q/best-record g (assoc a :venue venue))
                     label (case (:venue a) "home" "home " "away" "away " "")]
                 {:text (str (format "Best %srecords:\n" label)
                             (->> rows
                                  (map-indexed (fn [i r]
                                                 (format "%d. %s - %.1f%% (%dW %dD %dL of %d)"
                                                         (inc i) (:team r) (double (:win-rate r))
                                                         (:wins r) (:draws r) (:losses r) (:matches r))))
                                  (str/join "\n")))
                  :data rows}))}

   {:name "list_competitions"
    :description "List the competitions available in the knowledge graph with season coverage and match counts."
    :inputSchema {:type "object" :properties {}}
    :handler (fn [g _]
               (let [cs (q/list-competitions g)]
                 {:text (fmt/competitions-block cs) :data cs}))}])

(def ^:private tools-by-name (into {} (map (juxt :name identity) tools)))

(defn tool-list-payload
  "The public {:name :description :inputSchema} view returned by tools/list."
  []
  (mapv #(select-keys % [:name :description :inputSchema]) tools))

;; ---------------------------------------------------------------------------
;; Tool invocation
;; ---------------------------------------------------------------------------

(defn call-tool
  "Invoke a tool by name with an argument map (string OR keyword keys).
   Returns an MCP tool result map. Errors are reported via :isError true so the
   model can recover rather than the whole request failing."
  [tool-name raw-args]
  (if-let [tool (get tools-by-name tool-name)]
    (try
      (let [args (reduce-kv (fn [m k v] (assoc m (keyword k) v)) {} (or raw-args {}))
            {:keys [text data]} ((:handler tool) (data/graph) args)]
        {:content [{:type "text" :text text}]
         :structuredContent {:result data}
         :isError false})
      (catch Exception e
        {:content [{:type "text" :text (str "Error running tool '" tool-name "': " (.getMessage e))}]
         :isError true}))
    {:content [{:type "text" :text (str "Unknown tool: " tool-name)}]
     :isError true}))

;; ---------------------------------------------------------------------------
;; JSON-RPC request handling (pure)
;; ---------------------------------------------------------------------------

(defn handle-request
  "Pure JSON-RPC dispatch. Returns a response map, or nil for notifications
   (requests without an :id) which must not be answered."
  [{:keys [id method params] :as req}]
  (let [ok  (fn [result] {:jsonrpc "2.0" :id id :result result})
        err (fn [code msg] {:jsonrpc "2.0" :id id :error {:code code :message msg}})]
    (case method
      "initialize"
      (ok {:protocolVersion protocol-version
           :capabilities {:tools {:listChanged false}}
           :serverInfo server-info})

      "ping" (ok {})

      "tools/list" (ok {:tools (tool-list-payload)})

      "tools/call"
      (let [{:keys [name arguments]} params]
        (ok (call-tool name arguments)))

      ;; Notifications (initialized, cancelled, ...) carry no id -> no response.
      (if (nil? id)
        nil
        (err -32601 (str "Method not found: " method))))))

;; ---------------------------------------------------------------------------
;; stdio transport (line-delimited JSON, one message per line)
;; ---------------------------------------------------------------------------

(defn serve!
  "Run the stdio server loop, reading newline-delimited JSON-RPC messages from
   `in` and writing responses to `out`. Blocks until EOF."
  [^BufferedReader in out]
  (binding [*out* out]
    (loop []
      (when-let [line (.readLine in)]
        (when-not (str/blank? line)
          (let [resp (try
                       (handle-request (json/read-str line :key-fn keyword))
                       (catch Exception e
                         {:jsonrpc "2.0" :id nil
                          :error {:code -32700 :message (str "Parse error: " (.getMessage e))}}))]
            (when resp
              (println (json/write-str resp))
              (flush))))
        (recur)))))
